### PR TITLE
miniupnpd: add force_forwarding option support

### PR DESCRIPTION
**Pull request by ptpt52**
This option replace the 'ext_ip_reserved_ignore', and replace the
patch with new one

as addr_is_reserved() behavior should not be changed, so drop the
ext_ip_reserved_ignore

use a force_forwarding option to make the port forwarding force to
work when the router is behind NAT

by default force_forwarding is set to 0(disabled), where change
nothing with the original behavior, and users can turn it on
according to their wishes

related: https://github.com/openwrt/packages/pull/16011

### CODE REVIEW

**Comment by ptpt52** [comment]
for use case:
1. router get private IP but it is full cone nat environment, the force forwarding UPNP redirect works.

**Comment by ptpt52** [comment]
Theoretically and technically, all IPs are equal. Don't make any difference because it is a private IP. Under the understanding of this concept, this option is useful

**Comment by champtar** [comment]
You need to refresh the patches (make package/miniupnpd/refresh)

File: net/miniupnpd/patches/301-options-force_forwarding-support.patch [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
++++ b/upnpsoap.c
+@@ -348,7 +348,11 @@ GetExternalIPAddress(struct upnphttp * h, const char * action, const char * ns)
+ 			ext_ip_addr[0] = '\0';
+ 		} else if (addr_is_reserved(&addr)) {
+ 			syslog(LOG_NOTICE, "private/reserved address %s is not suitable for external IP", ext_ip_addr);
+-			ext_ip_addr[0] = '\0';
++			if (!GETFLAG(FORCEFORWARDINGMASK)) {
++				ext_ip_addr[0] = '\0';
++			} else {
++				syslog(LOG_NOTICE, "force_forwarding enable, private/reserved address %s used as external IP", ext_ip_addr);
```
Comment (stangri) [inline net/miniupnpd/patches/301-options-force_forwarding-support.patch:202]: If I understand the concept correctly, `force_forwarding enabled, private/reserved address %s used as external IP` is a correct way of saying this.

**Review by stangri — CHANGES_REQUESTED** [review]

**Comment by stangri** [comment]
> Theoretically and technically, all IPs are equal. Don't make any difference even if it is a private IP. Under this concept, this option is useful

Since this came up before, would it be possible to submit this upstream so that no modifications are needed for OpenWrt package?

Also, including @BKPepe as he has merged some PRs for this before.

**Review by BKPepe — CHANGES_REQUESTED** [review]
In your latest changes, you tried to upstream something, which was refused to be merged by upstream as you were changing some logics, etc. It was clearly hacked, which was merged here, and your warnings, errors were cryptic. Nobody knows what they meant. (= https://github.com/openwrt/packages/issues/15258 )

My response from the previous PR with the same changes didn't change:
https://github.com/openwrt/packages/pull/16011#pullrequestreview-698143667

TL;DR: We can not carry patches that are refused to be part of upstream and are causing several (even security) issues.  (= https://github.com/miniupnp/miniupnp/issues/556#issuecomment-907191205 ). Cooperate with upstream to get it right.

In the meantime, NAK.

**Comment by BKPepe** [comment]
This was not sent to upstream and because of that, I'm closing it.

**Comment by rmilecki** [comment]
Our attempts to upstream this feature have failed, see:
* https://github.com/miniupnp/miniupnp/pull/511
* https://github.com/miniupnp/miniupnp/pull/741

It seems however a similar option (`ignore_private_ip`) was added in the upstream by https://github.com/miniupnp/miniupnp/pull/813 and is present in the 2.3.9 release.
